#### Setup

*Connecting to the PostgreSQL database and loading the `ipython-sql` extension so SQL queries can run directly inside the notebook (via the `%%sql` magic).*

In [1]:
import pandas as pd
from sqlalchemy import create_engine
%load_ext sql

In [2]:
engine=create_engine('postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera')
connection=engine.connect()
print("Connected:", connection.closed == False)

Connected: True


In [3]:
%sql postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera

'Connected: postgres@medika_sejahtera'

#### Business Question

1. What reason code most frequently causes product returns?
2. Which products and product categories are returned most often (by frequency)?
3. What is the total financial loss from returns across 2024, and which products/categories account for the largest share?
4. Which customer type (healthcare institution type) has the most returns?
5. Which region records the highest number of return cases?
6. What does the return trend look like across 2024 — is there a monthly or seasonal pattern?

# 1. Product & Reason Analysis

*Baseline: total count and share (%) of returns by reason code, across the full dataset.*

In [4]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT (*) AS total_record
FROM
    return_cleaned
GROUP BY
    return_reason_cleaned
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_record
Salah Input Sistem,28
Salah Pesan Customer,20
Barang Rusak,7
Kadaluarsa,4


In [27]:
%%sql
WITH total AS
(
    SELECT
        COUNT(*) AS total_records
    FROM
        return_cleaned
)

SELECT
    rc.return_reason_cleaned,
    COUNT(*) AS total_record,
    tl.total_records AS total_data,
    ROUND((COUNT(*)::numeric / NULLIF(tl.total_records,0)::numeric) * 100,2) AS rate
FROM
    return_cleaned rc
CROSS JOIN
    total tl
GROUP BY
    rc.return_reason_cleaned,
    total_data
ORDER BY
    total_record DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_record,total_data,rate
Salah Input Sistem,28,59,47.46
Salah Pesan Customer,20,59,33.90
Barang Rusak,7,59,11.86
Kadaluarsa,4,59,6.78


**Screening**

System Input Error (Salah Input Sistem) (SIS, 47.46%) and Customer Ordering Error (Salah Pesan Customer) (SPC, 33.90%) are both major contributors, together accounting for 81.36% of all cases. But they come from two different failure points — one from an internal process, the other from the customer's side when placing an order — so they need two separate fixes, not a single blanket solution.

#### Screening 1D

(baseline per reason code, then broken down by region / quarter / category / customer type / product)

##### *Reason code breakdown by quarter.*

In [35]:
%%sql

WITH reason_per_quarter AS (
    SELECT
        return_reason_cleaned,
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        COUNT(*) AS total_record_filter
    FROM 
        return_cleaned
    GROUP BY 
        return_reason_cleaned, 
        quarter
)
SELECT
    return_reason_cleaned,
    quarter,
    total_record_filter,
    SUM(total_record_filter) OVER (PARTITION BY quarter) AS total_per_quarter,
    ROUND(
        total_record_filter * 100.0 / SUM(total_record_filter) OVER (PARTITION BY quarter),
        2
    ) AS row_pct
FROM 
    reason_per_quarter
ORDER BY 
    return_reason_cleaned,
    quarter, 
    row_pct DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
14 rows affected.


return_reason_cleaned,quarter,total_record_filter,total_per_quarter,row_pct
Barang Rusak,Q1,3,8,37.50
Barang Rusak,Q2,2,15,13.33
Barang Rusak,Q3,2,14,14.29
Kadaluarsa,Q1,1,8,12.50
Kadaluarsa,Q3,2,14,14.29
Kadaluarsa,Q4,1,22,4.55
Salah Input Sistem,Q1,2,8,25.00
Salah Input Sistem,Q2,10,15,66.67
Salah Input Sistem,Q3,6,14,42.86
Salah Input Sistem,Q4,10,22,45.45


In [36]:
%%sql

SELECT
    CASE
        WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 10 AND 12 THEN 'Q4'
    END AS quarter,
    COUNT(*) AS total_data
FROM 
    invoice_data
GROUP BY 
    quarter
ORDER BY 
    quarter

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_data
Q1,149
Q2,171
Q3,179
Q4,204


*Return rate relative to invoice volume per quarter, focused on SIS and SPC, with quarter-over-quarter growth — this is the query behind the screening findings below.*

In [37]:
%%sql
WITH return_by_reason_quarter AS (
    SELECT
        return_reason_cleaned,
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        COUNT(*) AS return_count
    FROM return_cleaned
    GROUP BY return_reason_cleaned, quarter
),
invoice_by_quarter AS (
    SELECT
        CASE
            WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM invoice_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        COUNT(*) AS invoice_count
    FROM invoice_data
    GROUP BY quarter
),
rate_by_quarter AS (
    SELECT
        rq.return_reason_cleaned,
        rq.quarter,
        rq.return_count,
        iq.invoice_count,
        ROUND(
            (rq.return_count::numeric * 100) / NULLIF(iq.invoice_count, 0)::numeric,
            2
        ) AS rate_pct
    FROM return_by_reason_quarter rq
    JOIN invoice_by_quarter iq ON rq.quarter = iq.quarter
)
SELECT
    return_reason_cleaned,
    quarter,
    return_count,
    invoice_count,
    rate_pct,
    ROUND(
        (rate_pct - LAG(rate_pct) OVER (PARTITION BY return_reason_cleaned ORDER BY quarter))
        / NULLIF(LAG(rate_pct) OVER (PARTITION BY return_reason_cleaned ORDER BY quarter), 0)
        * 100,
        2
    ) AS growth_pct_qoq
FROM rate_by_quarter
WHERE return_reason_cleaned IN ('Salah Input Sistem', 'Salah Pesan Customer')
ORDER BY return_reason_cleaned, quarter;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
8 rows affected.


return_reason_cleaned,quarter,return_count,invoice_count,rate_pct,growth_pct_qoq
Salah Input Sistem,Q1,2,149,1.34,None
Salah Input Sistem,Q2,10,171,5.85,336.57
Salah Input Sistem,Q3,6,179,3.35,-42.74
Salah Input Sistem,Q4,10,204,4.90,46.27
Salah Pesan Customer,Q1,2,149,1.34,None
Salah Pesan Customer,Q2,3,171,1.75,30.60
Salah Pesan Customer,Q3,4,179,2.23,27.43
Salah Pesan Customer,Q4,11,204,5.39,141.70


**Screening**  


Q2 (n=15) — SIS spikes (66.67%, +19.21 pts), SPC dips (20%, -13.90 pts) → worth digging into

Q4 (n=22) — SPC spikes (50%, +16.10 pts), SIS roughly at baseline (45.45%, -2.01 pts, negligible) → worth digging into

Q3 (n=14) — all reasons close to baseline → clean screening

Q1 (n=8, individual cells n=1-3) — large deviations on paper (SIS -22.46, Damaged Goods +25.64), but cell sizes are too small → minor observation 

##### *Reason code breakdown by region.*

In [63]:
%%sql

WITH reason_count AS (
    SELECT 
        region_cleaned,
        return_reason_cleaned,
        COUNT(*) AS n
    FROM 
        return_cleaned
    GROUP BY  
        region_cleaned, 
        return_reason_cleaned
)
SELECT 
    region_cleaned,
    return_reason_cleaned,
    n,
    SUM(n) OVER (PARTITION BY region_cleaned) AS total_per_region,
    ROUND(
        n * 100.0 / SUM(n) OVER (PARTITION BY region_cleaned), 
        2
    ) AS row_pct
FROM 
    reason_count
ORDER BY
    region_cleaned,
    row_pct DESC


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
25 rows affected.


region_cleaned,return_reason_cleaned,n,total_per_region,row_pct
Bekasi,Salah Input Sistem,5,10,50.00
Bekasi,Salah Pesan Customer,4,10,40.00
Bekasi,Kadaluarsa,1,10,10.00
Bogor,Salah Input Sistem,2,4,50.00
Bogor,Salah Pesan Customer,2,4,50.00
Depok,Kadaluarsa,1,3,33.33
Depok,Barang Rusak,1,3,33.33
Depok,Salah Input Sistem,1,3,33.33
Jakarta Barat,Salah Input Sistem,8,13,61.54
Jakarta Barat,Salah Pesan Customer,3,13,23.08


**Screening**  
West Jakarta (n=13) — SIS stands out (61.54%, +14.08 pts), SPC low (23.08%, -10.82 pts) → worth digging into

South Jakarta (n=10) — both SIS (50%, +2.54 pts) and SPC (50%, +16.10 pts) above baseline → worth digging into

East Jakarta (n=6) — SPC stands out (50%, +16.10 pts) → worth digging into (small n)

Tangerang (n=5, Damaged Goods cell n=3) — Damaged Goods high (60%, baseline 11.86%, +48.14 pts) → worth digging into (very small n, treat with caution)

Bekasi (n=10) — all reasons close to baseline → clean screening

Central Jakarta, Bogor, Depok, North Jakarta (n=3-5 per region, individual cells n=1-2) → n too small to draw conclusions, minor observation only

##### *Reason code breakdown by category.*

In [65]:
%%sql

WITH category_count AS (
    SELECT
        kategori,
        return_reason_cleaned,
        COUNT (*) AS total_records
    FROM
        return_cleaned
    GROUP BY
        kategori,
        return_reason_cleaned
)


SELECT
    kategori,
    return_reason_cleaned,
    total_records,
    SUM(total_records) OVER (PARTITION BY kategori) AS total_per_category,
    ROUND (total_records * 100.0 / SUM(total_records) OVER (PARTITION BY kategori),2) AS row_pct
FROM
    category_count
ORDER BY
    kategori,
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
17 rows affected.


kategori,return_reason_cleaned,total_records,total_per_category,row_pct
Bedah & Tindakan,Salah Input Sistem,5,11,45.45
Bedah & Tindakan,Salah Pesan Customer,5,11,45.45
Bedah & Tindakan,Kadaluarsa,1,11,9.09
Consumable,Salah Input Sistem,6,13,46.15
Consumable,Salah Pesan Customer,4,13,30.77
Consumable,Kadaluarsa,2,13,15.38
Consumable,Barang Rusak,1,13,7.69
Diagnostik,Salah Input Sistem,9,16,56.25
Diagnostik,Salah Pesan Customer,4,16,25.00
Diagnostik,Barang Rusak,2,16,12.50


**Screening**  
Diagnostics — SIS stands out (56.25%, baseline 47.46%, +8.79 pts), SPC low (25%, baseline 33.90%, -8.90 pts) → worth digging into

Laboratory — SIS stands out (62.5%, baseline 47.46%, +15.04 pts), SPC low (25%, baseline 33.90%, -8.90 pts) → worth digging into (n=8, small — read alongside Diagnostics as a related category, not on its own)

Surgery & Procedures (Bedah & Tindakan) — SPC stands out (45.45%, baseline 33.90%, +11.55 pts), SIS close to baseline (45.45% vs. 47.46%, -2.01 pts, negligible) → worth digging into

Nursing & Inpatient Care (Perawatan & Rawat Inap) — SPC stands out (45.45%, baseline 33.90%, +11.55 pts), SIS low (27.27%, baseline 47.46%, -20.19 pts) → worth digging into

##### *Reason code breakdown by customer type.*

In [66]:
%%sql

WITH customer_type_count AS (
    SELECT
            customer_type_cleaned,
            return_reason_cleaned,
            COUNT (*) AS total_records
        FROM
            return_cleaned
        GROUP BY
            customer_type_cleaned,
            return_reason_cleaned

)

SELECT
    customer_type_cleaned,
    return_reason_cleaned,
    total_records,
    SUM(total_records) OVER (PARTITION BY customer_type_cleaned) AS total_per_customer_type,
    ROUND (total_records * 100.0 / SUM(total_records) OVER (PARTITION BY customer_type_cleaned),2) AS row_pct
FROM
    customer_type_count
ORDER BY
    customer_type_cleaned,
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
12 rows affected.


customer_type_cleaned,return_reason_cleaned,total_records,total_per_customer_type,row_pct
Klinik,Salah Input Sistem,9,19,47.37
Klinik,Salah Pesan Customer,8,19,42.11
Klinik,Barang Rusak,1,19,5.26
Klinik,Kadaluarsa,1,19,5.26
RS Pemerintah,Salah Input Sistem,8,13,61.54
RS Pemerintah,Barang Rusak,2,13,15.38
RS Pemerintah,Salah Pesan Customer,2,13,15.38
RS Pemerintah,Kadaluarsa,1,13,7.69
RS Swasta,Salah Input Sistem,11,27,40.74
RS Swasta,Salah Pesan Customer,10,27,37.04


**Screening**  
Government Hospital (RS Pemerintah) — SIS stands out (61.54%, baseline 47.46%, +14.08 pts), SPC low (15.38%, baseline 33.90%) → worth digging into

Clinic — SPC deviates (42.11%, baseline 33.90%, +8.21 pts), SIS at baseline (47.37% vs. 47.46%, negligible) → worth digging into (SPC specifically)

Private Hospital (RS Swasta) — SIS (40.74%, -6.72 pts) and SPC (37.04%, +3.14 pts), both close to baseline → clean screening, no meaningful deviation

##### *Reason code breakdown at the individual product (SKU) level.*

In [67]:
%%sql

WITH product_count AS (
    SELECT
            sku_name,
            return_reason_cleaned,
            COUNT (*) AS total_records
        FROM
            return_cleaned
        GROUP BY
            sku_name,
            return_reason_cleaned

)

SELECT
    sku_name,
    return_reason_cleaned,
    total_records,
    SUM(total_records) OVER (PARTITION BY sku_name) AS total_per_product,
    ROUND (total_records * 100.0 / SUM(total_records) OVER (PARTITION BY sku_name),2) AS row_pct
FROM
    product_count
ORDER BY
    sku_name,
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
50 rows affected.


sku_name,return_reason_cleaned,total_records,total_per_product,row_pct
Baju Pasien Dewasa,Barang Rusak,1,2,50.00
Baju Pasien Dewasa,Salah Input Sistem,1,2,50.00
Bantal Angin Anti Decubitus,Salah Pesan Customer,1,1,100.00
Bedpan Plastik Dewasa,Salah Input Sistem,1,1,100.00
Benang Catgut 2-0,Salah Input Sistem,2,3,66.67
Benang Catgut 2-0,Salah Pesan Customer,1,3,33.33
Duk Operasi 80x90cm,Kadaluarsa,1,2,50.00
Duk Operasi 80x90cm,Salah Pesan Customer,1,2,50.00
Glukometer Set,Salah Input Sistem,2,2,100.00
Gunting Bedah Bengkok 14cm,Salah Input Sistem,1,2,50.00


**Screening**  
Analysis at the product (SKU) level wasn't pursued further — the granularity is too fine. Out of 48 unique products, the maximum number of return cases for any single product is only 3, and most have just 1. The more meaningful pattern is already captured at the category level (Diagnostics and Laboratory show a concentration of SIS; Surgery & Procedures and Nursing & Inpatient Care show a concentration of SPC).

#### Cross-Tab / Confounding Check

(West Jakarta × customer type — checking whether these findings are actually related)

*Customer type breakdown, filtered to West Jakarta only.*

In [68]:
%%sql

SELECT
    customer_type_cleaned,
    COUNT (*) AS total_records
FROM
    return_cleaned
WHERE
    region_cleaned = 'Jakarta Barat'
GROUP BY
    customer_type_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_records
RS Swasta,6
Klinik,4
RS Pemerintah,3


**Screening**  
n and row_pct match West Jakarta's numbers exactly (8/13) → verified in Section 2; result: independent (not a duplicate signal)

#### Category & Product

In [70]:
%%sql

WITH
    kategori_total AS (
        SELECT
            COUNT(*) AS total_loss
        FROM
            return_cleaned
    )

SELECT
    kategori,
    COUNT(*) AS total_loss_filter,
    kt.total_loss,
    ROUND((COUNT(*)::numeric / NULLIF(kt.total_loss,0)::numeric) * 100,2) AS row_pct
FROM
    return_cleaned rc
CROSS JOIN 
    kategori_total kt
GROUP BY
    kategori,
    kt.total_loss
ORDER BY
    total_loss_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_loss_filter,total_loss,row_pct
Diagnostik,16,59,27.12
Consumable,13,59,22.03
Perawatan & Rawat Inap,11,59,18.64
Bedah & Tindakan,11,59,18.64
Laboratorium,8,59,13.56


**Screening**  
Range of 8-16 cases, with a shallow gap between categories (the largest difference across all 5 categories is just 8 points) → the distribution is fairly even, with no category dominating drastically.

*The product (SKU) level, top 10 by return count.*

In [71]:
%%sql

WITH
    product_total AS (
        SELECT
            COUNT(*) AS total_loss
        FROM
            return_cleaned
    )

SELECT
    sku_name,
    COUNT(*) AS total_loss_filter,
    pt.total_loss,
    ROUND((COUNT(*)::numeric / NULLIF(pt.total_loss,0)::numeric) * 100,2) AS ratio
FROM
    return_cleaned rc
CROSS JOIN 
    product_total pt
GROUP BY
    sku_name,
    pt.total_loss
ORDER BY
    total_loss_filter DESC
LIMIT 10

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


sku_name,total_loss_filter,total_loss,ratio
Syringe 1ml,3,59,5.08
Urine Stick 10 Parameter,3,59,5.08
Infus Set Anak,3,59,5.08
Selimut Pasien Rumah Sakit,3,59,5.08
Benang Catgut 2-0,3,59,5.08
IV Catheter 22G,2,59,3.39
Gunting Bedah Bengkok 14cm,2,59,3.39
NGT 16Fr,2,59,3.39
Tensimeter Manual Dewasa,2,59,3.39
Baju Pasien Dewasa,2,59,3.39


**Screening**  
The maximum count for any single product is only 3 (out of 48 unique SKUs), and most have just 1-2. The granularity is too fine to draw any conclusion about the distribution pattern — whether "dominant" or "even" — since one extra case could shift the ranking significantly. The product level isn't analyzed further; the category level is sufficient to answer BQ2.

**Verification:**  
**Trigger:** West Jakarta (region, SIS 8/13) and Government Hospital (customer type, SIS 8/13) show identical numbers → checked whether this is actually the same signal counted twice.

**Cross-tab:** West Jakarta × customer type → Private Hospital (6), Clinic (4), Government Hospital (3)

**Conclusion:** Government Hospital only accounts for 3 of the 13 cases in West Jakarta (not the majority) → these two findings are INDEPENDENT, not one overlapping signal. They can be reported as two separate findings.

#### Findings

System Input Error (Salah Input Sistem, SIS) is the most dominant reason code (47.46%), concentrated especially in West Jakarta (61.54%, +14.08 pts above baseline) and the Diagnostics/Laboratory categories (56-62%), with a sharp spike in Q2 (66.67%). The SIS rate relative to total transactions jumped 336% from Q1 to Q2 — far outpacing the transaction volume growth over the same period (14.8%) — which points to a cause beyond simple order volume; the specific root cause (a staffing or process change) can't be pinned down from the data available. Left unaddressed, this pattern will likely keep recurring above baseline, adding to the operational load of processing returns — especially for the team handling Diagnostics/Lab products in West Jakarta. Recommend investigating system change logs, input SOPs, or staffing around Q2, prioritizing the region and categories identified above.

Customer Ordering Error (Salah Pesan Customer, SPC) is the second-largest reason code (33.90%), with a different pattern from SIS — instead of a spike at one point in time, it's a trend that keeps worsening across the year: the SPC rate relative to total transactions rose consistently every quarter (1.34% → 1.75% → 2.23% → 5.39%), growing far faster than transaction volume over the same period (141% vs. 14% from Q3 to Q4). SPC is also concentrated in the Surgery & Procedures and Nursing & Inpatient Care categories (+11.55 pts above baseline each) and in the Clinic customer type (+8.21 pts) — likely tied to manual PO processes from customers that aren't automatically validated, though the specific cause behind this worsening trend can't be confirmed from the data available. At the regional level, SPC is also high in South Jakarta, but mixed in with an equally high SIS rate (50%/50%) — the relationship between the two hasn't been explored further. If left unaddressed, the return burden from SPC could keep growing faster than the business itself — a sign of a process problem, not just a natural consequence of more transactions. Recommend investigating the PO receiving/verification process, particularly for the Surgery & Procedures/Nursing & Inpatient Care categories and the Clinic customer type.

The distribution of returned product categories is fairly even (range of 8-16 cases), with no category dominating drastically; the product (SKU) level can't be analyzed further since the maximum count for any product is only 3 out of 48 unique SKUs. This even distribution actually reinforces the BQ1 finding — since returns aren't concentrated on specific products or categories, the root cause is unlikely to be product characteristics (e.g., a design or quality issue with a particular item), and more likely a process problem affecting all product lines equally — consistent with SIS (system input errors) and SPC (ordering mistakes from manual POs), both of which are process issues, not product issues. This means fixes targeted at specific products or categories (e.g., extra quality control for 1-2 SKUs) are unlikely to meaningfully reduce total return cases, since the problem is spread across every category. Recommend keeping the improvement focus on process (system input validation, PO verification) as recommended in BQ1, rather than product-specific interventions.

# 2. Financial Impact

*Total financial loss and average loss per return case, across the full dataset.*

In [64]:
%%sql

SELECT
    SUM(value_return) AS total_loss,
    AVG(value_return) AS avg
FROM
    return_cleaned


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_loss,avg
171506500.0,2906889.8305084747


**Screening**  
Total financial loss from returns across 2024 reached IDR 171,506,500 from 59 return cases, averaging IDR 2,906,890 per case.

*Financial loss and case count broken down by category.*

In [72]:
%%sql

SELECT
    kategori,
    SUM(value_return) AS total_loss,
    COUNT(value_return) AS total_record
FROM
    return_cleaned
GROUP BY
    kategori
ORDER BY
    total_loss DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_loss,total_record
Diagnostik,66811000.0,16
Bedah & Tindakan,53336500.0,11
Perawatan & Rawat Inap,22706500.0,11
Consumable,21474000.0,13
Laboratorium,7178500.0,8


In [74]:
%%sql

WITH
    kategori_total AS (
        SELECT
            SUM(value_return) AS total_loss,
            COUNT(*) AS total_records
        FROM
            return_cleaned
    )

SELECT
    kategori,
    SUM(value_return) AS total_loss_filter,
    kt.total_loss,
    ROUND((SUM(value_return)::numeric / NULLIF(kt.total_loss,0)::numeric) * 100,2) AS loss_pct,
    ROUND(AVG(value_return):: numeric,2) AS average_loss_filter,
    COUNT(value_return) AS total_record_filter,
    kt.total_records,
    ROUND((COUNT(*)::numeric / NULLIF(total_records,0)::numeric) * 100,2) AS row_pct
FROM
    return_cleaned rc
CROSS JOIN 
    kategori_total kt
GROUP BY
    kategori,
    kt.total_loss,
    kt.total_records
ORDER BY
    total_loss_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_loss_filter,total_loss,loss_pct,average_loss_filter,total_record_filter,total_records,row_pct
Diagnostik,66811000.0,171506500.0,38.96,4175687.50,16,59,27.12
Bedah & Tindakan,53336500.0,171506500.0,31.10,4848772.73,11,59,18.64
Perawatan & Rawat Inap,22706500.0,171506500.0,13.24,2064227.27,11,59,18.64
Consumable,21474000.0,171506500.0,12.52,1651846.15,13,59,22.03
Laboratorium,7178500.0,171506500.0,4.19,897312.50,8,59,13.56


**Screening**  
Diagnostics (n=16, 27.12% of cases) — 38.96% of total loss (IDR 66.8M), gap +11.84 pts vs. its case share, avg IDR 4,175,688/case → disproportionately costly, driven by volume (highest case count) → worth digging into

Surgery & Procedures (n=11, 18.64% of cases) — 31.10% of total loss (IDR 53.3M), gap +12.46 pts (the LARGEST gap of any category), avg IDR 4,848,773/case (the HIGHEST) → the MOST disproportionately costly, driven by value per case rather than volume → worth digging into

Nursing & Inpatient Care (n=11, 18.64% of cases) — 13.24% of total loss (IDR 22.7M), gap -5.40 pts, avg IDR 2,064,227/case → close to proportional → reasonably clean screening

Consumables (n=13, 22.03% of cases) — 12.52% of total loss (IDR 21.5M), gap -9.51 pts, avg IDR 1,651,846/case → disproportionately cheap (many cases, low value) → not a loss priority

Laboratory (n=8, 13.56% of cases) — 4.19% of total loss (IDR 7.2M), gap -9.37 pts, avg IDR 897,313/case (the LOWEST) → disproportionately cheap → not a loss priority

*Drilling into individual Diagnostics return lines — joining with invoice and return-quantity detail to identify which specific orders drive the loss.*

In [12]:
%%sql

WITH 
month AS (
    SELECT
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        SUM(value_return) AS total_loss,
        COUNT(*) AS total_records
    FROM
        return_cleaned
    GROUP BY
        quarter
),
month_filter AS (
    SELECT
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        SUM(value_return) AS total_loss,
        COUNT(*) AS total_record_filter,
        AVG(value_return) AS average_loss_filter
    FROM 
        return_cleaned
    WHERE 
        kategori = 'Diagnostik'
    GROUP BY 
        quarter
)


SELECT
    m.quarter,
    COALESCE(mf.total_loss,0) AS total_loss,
    m.total_loss AS total_loss_all,
    ROUND((COALESCE(mf.total_loss,0) * 1.0 / NULLIF(m.total_loss,0) * 100)::numeric,2) AS rate,
    COALESCE(mf.total_record_filter,0) AS total_record_filter,
    COALESCE(m.total_records,0) AS total_records,
    COALESCE(mf.average_loss_filter,0) AS average_loss_filter
FROM month m
LEFT JOIN month_filter mf 
    ON m.quarter = mf.quarter
ORDER BY 
    m.quarter

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_loss,total_loss_all,rate,total_record_filter,total_records,average_loss_filter
Q1,3510000.0,30449500.0,11.53,3,8,1170000.0
Q2,33395000.0,69647000.0,47.95,4,15,8348750.0
Q3,25298000.0,34531500.0,73.26,4,14,6324500.0
Q4,4608000.0,36878500.0,12.50,5,22,921600.0


**Findings**
Of the 4 quarters, only Q4 is reliable enough for comparison (observed 5 cases, expected 5.97 cases, narrowly above the threshold). The result shows Diagnostic-category loss in Q4 falls well below expectation (deviation of -IDR 9,758,168; rate 12.50% vs. the 38.96% baseline). In contrast, Q3 shows a high positive deviation (+IDR 11,846,116, 73.26% rate) — but this quarter is not reliable, as its case count (4) falls below the threshold.

In [31]:
%%sql

WITH kategori_reason AS
(
    SELECT
        return_reason_cleaned,
        COUNT(*) AS total_records_filter,
        SUM(value_return) AS total_loss_filter,
        AVG(value_return) AS average_loss
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        return_reason_cleaned
)

SELECT
    rc.return_reason_cleaned,
    kr.total_loss_filter AS total_loss_filter,
    SUM(rc.value_return) AS total_loss,
    ROUND((COALESCE(kr.total_loss_filter,0) * 1.0 / NULLIF(SUM(rc.value_return), 0)*100)::numeric, 2) AS rate,
    ROUND(kr.average_loss::numeric, 2) AS average_loss_filter,
    kr.total_records_filter AS total_records_filter,
    COUNT(*) AS total_records
    
FROM
    return_cleaned rc
LEFT JOIN
    kategori_reason kr
        ON rc.return_reason_cleaned = kr.return_reason_cleaned
GROUP BY
    rc.return_reason_cleaned,
    total_loss_filter,
    total_records_filter,
    average_loss_filter
ORDER BY
    total_loss_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_loss_filter,total_loss,rate,average_loss_filter,total_records_filter,total_records
Salah Input Sistem,26842000.0,66634500.0,40.28,2982444.44,9,28
Salah Pesan Customer,19761000.0,70452500.0,28.05,4940250.00,4,20
Barang Rusak,17447000.0,29144000.0,59.86,8723500.00,2,7
Kadaluarsa,2761000.0,5275500.0,52.34,2761000.00,1,4


**Findings**  
System Input Error is the only reliable reason for Diagnostic-category loss (observed 9 cases vs. expected 7.59, both above threshold), with a loss of IDR 26,842,000 and a rate of 40.28% — slightly above (+1.32 points) the 38.96% baseline, with a deviation of +IDR 881,199. Customer Order Error and Damaged Goods are both unreliable, though from different angles: Customer Order Error has a passing expected count (5.42) but only 4 observed cases, while Damaged Goods records the highest rate among all reasons (59.86%) but is based on just 2 of 7 cases — both too small to support a conclusion.

In [33]:
%%sql

WITH kategori_customer AS
(
    SELECT
        customer_type_cleaned,
        COUNT(*) AS total_records,
        SUM(value_return) AS total_loss,
        AVG(value_return) AS average_loss
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        customer_type_cleaned 
)

SELECT
    rc.customer_type_cleaned,
    kc.total_loss AS total_loss_filter,
    SUM(rc.value_return) AS total_loss,
    ROUND((COALESCE(kc.total_loss,0) * 1.0 / NULLIF(SUM(rc.value_return), 0)*100)::numeric, 2) AS rate,
    ROUND(kc.average_loss::numeric, 2) AS average_loss_filter,
    kc.total_records AS total_records_filter,
    COUNT(*) AS total_records
FROM
    return_cleaned rc
LEFT JOIN
    kategori_customer kc
        ON rc.customer_type_cleaned = kc.customer_type_cleaned
GROUP BY
    rc.customer_type_cleaned,
    total_loss_filter,
    total_records_filter,
    average_loss_filter
ORDER BY
    total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_loss_filter,total_loss,rate,average_loss_filter,total_records_filter,total_records
Klinik,29225000.0,52240500.0,55.94,4175000.00,7,19
RS Swasta,18305000.0,76510000.0,23.92,3050833.33,6,27
RS Pemerintah,19281000.0,42756000.0,45.10,6427000.00,3,13


**Findings**  
Clinics are the only customer type reliable across all aspects for Diagnostic-category loss — observed 7 cases vs. expected 5.15 cases (both above threshold), with a rate of 55.94% (+16.98 points above the 38.96% baseline) and a deviation of +IDR 8,872,101. Private hospitals are also reliable by case count, but their rate falls below baseline (23.92%). In contrast, public hospitals record the highest rate (45.10%) but are not reliable at all, with a base of only 3 of 13 cases.

In [35]:
%%sql

WITH kategori_region AS
(
    SELECT
        region_cleaned,
        COUNT(*) AS total_records,
        SUM(value_return) AS total_loss,
        AVG(value_return) AS average_loss
    FROM
        return_cleaned
    WHERE
        kategori = 'Diagnostik'
    GROUP BY
        region_cleaned
)

SELECT
    rc.region_cleaned,
    kr.total_loss AS total_loss_filter,
    SUM(rc.value_return) AS total_loss,
    ROUND((COALESCE(kr.total_loss,0) * 1.0 / NULLIF(SUM(rc.value_return), 0)*100)::numeric, 2) AS rate,
    ROUND((COALESCE(kr.average_loss,0))::numeric, 2) AS average_loss_filter,
    kr.total_records AS total_records_filter,
    COUNT(*) AS total_records
FROM
    return_cleaned rc
LEFT JOIN
    kategori_region kr
        ON rc.region_cleaned = kr.region_cleaned
GROUP BY
    rc.region_cleaned,
    total_records_filter,
    total_loss_filter,
    average_loss_filter
ORDER BY
    total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_loss_filter,total_loss,rate,average_loss_filter,total_records_filter,total_records
Bogor,None,21657000.0,0.00,0.00,None,4
Jakarta Timur,None,4650500.0,0.00,0.00,None,6
Jakarta Barat,19953000.0,33920000.0,58.82,2850428.57,7,13
Jakarta Selatan,7182000.0,11695500.0,61.41,2394000.00,3,10
Bekasi,18734000.0,36964500.0,50.68,9367000.00,2,10
Depok,1908000.0,3399500.0,56.13,1908000.00,1,3
Jakarta Pusat,1587000.0,19224000.0,8.26,1587000.00,1,3
Jakarta Utara,1468000.0,19318500.0,7.60,1468000.00,1,5
Tangerang,15979000.0,20677000.0,77.28,15979000.00,1,5


**Findings**  
No region meets the reliability threshold, primarily due to expected count — even Jakarta Barat, whose observed count passes (7 cases, above the threshold of 5), still fails on expected count (3.53). Exploratory note: Jakarta Barat records a 58.82% rate (+19.86 points above the 38.96% baseline, a deviation of +IDR 6,737,608), and Tangerang shows the highest rate among all regions (77.28%, a deviation of +IDR 7,923,223) — but Tangerang's base is only 1 of 5 cases, well below threshold. Neither figure supports a solid conclusion.

In [38]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'RS Swasta' ) AS RS_Swasta,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'Klinik' ) AS Klinik,
    COUNT(*) FILTER (WHERE customer_type_cleaned = 'RS Pemerintah' ) AS RS_Pemerintah,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    return_reason_cleaned
ORDER BY
    return_reason_cleaned

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,rs_swasta,klinik,rs_pemerintah,total_records
Barang Rusak,0,0,2,2
Kadaluarsa,1,0,0,1
Salah Input Sistem,4,4,1,9
Salah Pesan Customer,1,3,0,4


In [44]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3
    ) AS Q1,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6
    ) AS Q2,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9
    ) AS Q3,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12
    ) AS Q4,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    return_reason_cleaned
ORDER BY
    return_reason_cleaned


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,q1,q2,q3,q4,total_records
Barang Rusak,1,0,1,0,2
Kadaluarsa,0,0,1,0,1
Salah Input Sistem,1,3,2,3,9
Salah Pesan Customer,1,1,0,2,4


In [45]:
%%sql

SELECT
    customer_type_cleaned,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3
    ) AS Q1,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6
    ) AS Q2,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9
    ) AS Q3,
    COUNT(*) FILTER (
        WHERE EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12
    ) AS Q4,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    kategori = 'Diagnostik'
GROUP BY
    customer_type_cleaned
ORDER BY
    customer_type_cleaned


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,q1,q2,q3,q4,total_records
Klinik,2,2,0,3,7
RS Pemerintah,1,0,1,1,3
RS Swasta,0,2,3,1,6


**Findings**  
As with the frequency analysis (Section 1), cross-dimension exploration within the Diagnostic category (reason × customer type, customer type × quarter, reason × quarter) fails the minimum reliability threshold across all cells examined — a structural consequence of the Diagnostic category's small base (16 cases), which makes any two-dimensional split too granular to compare reliably. As such, no cross-dimension combination can support a solid financial loss conclusion.

---

In [48]:
%%sql

SELECT
    sku_name,
    SUM(value_return) AS total_loss,
    COUNT(value_return) AS total_record
FROM
    return_cleaned
GROUP BY
    sku_name
ORDER BY
    total_loss DESC
LIMIT 10

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


sku_name,total_loss,total_record
Rapid Test HbsAg,24174000.0,2
Urine Stick 10 Parameter,19189000.0,3
Benang Catgut 2-0,17976000.0,3
IV Catheter 22G,16345500.0,2
Handscoon Steril 6.5,15645000.0,1
Handscoon Steril 7.5,12801000.0,1
Kursi Roda Standar,9219000.0,1
Glukometer Set,7980000.0,2
Tiang Infus Stainless 5-Roda,6207500.0,1
Pulse Oximeter,5077000.0,2


In [56]:
%%sql

WITH
    sku_total AS (
        SELECT
            SUM(value_return) AS total_loss
        FROM
            return_cleaned
    )

SELECT
    sku_name,
    SUM(value_return) AS total_loss,
    sku.total_loss AS grand_total_loss,
    ROUND((SUM(value_return)::numeric / NULLIF(sku.total_loss,0)::numeric) * 100,2) AS rate,
    COUNT(value_return) AS total_record_filter,
    ROUND(AVG(value_return):: numeric,2) AS average_loss_filter
FROM
    return_cleaned rc
CROSS JOIN 
    sku_total sku
GROUP BY
    sku_name,
    sku.total_loss
ORDER BY
    total_loss DESC
LIMIT 10

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


sku_name,total_loss,grand_total_loss,rate,total_record_filter,average_loss_filter
Rapid Test HbsAg,24174000.0,171506500.0,14.10,2,12087000.00
Urine Stick 10 Parameter,19189000.0,171506500.0,11.19,3,6396333.33
Benang Catgut 2-0,17976000.0,171506500.0,10.48,3,5992000.00
IV Catheter 22G,16345500.0,171506500.0,9.53,2,8172750.00
Handscoon Steril 6.5,15645000.0,171506500.0,9.12,1,15645000.00
Handscoon Steril 7.5,12801000.0,171506500.0,7.46,1,12801000.00
Kursi Roda Standar,9219000.0,171506500.0,5.38,1,9219000.00
Glukometer Set,7980000.0,171506500.0,4.65,2,3990000.00
Tiang Infus Stainless 5-Roda,6207500.0,171506500.0,3.62,1,6207500.00
Pulse Oximeter,5077000.0,171506500.0,2.96,2,2538500.00


**Findings**  
The top 10 products by loss are each based on just 1–3 return cases (out of 59 total) — Rapid Test HbsAg records the highest loss (IDR 24,174,000, 14.10% of total, from 2 cases), followed by 10-Parameter Urine Stick (IDR 19,189,000, 11.19%, 3 cases) and 2-0 Catgut Suture (IDR 17,976,000, 10.48%, 3 cases). Given the very small case base per product, further breakdown (by time, region, or customer type) is not viable — this mirrors the pattern found at the product level in the frequency analysis (Section 1), where losses are thinly spread across many different products rather than concentrated in a few SKUs.

**Insight**  


**Insight**  


**Conclusion:** 
 

# 3. Who & Where

In [39]:
%%sql

SELECT
    customer_type_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
GROUP BY
    customer_type_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_records
RS Swasta,27
Klinik,19
RS Pemerintah,13


In [50]:
%%sql

WITH customer_type AS
(
    SELECT
        COUNT(*) AS total_records
    FROM
        return_cleaned
)

SELECT
    rc.customer_type_cleaned,
    COUNT(*) AS total_records__filter,
    ct.total_records AS total_records,
    ROUND((COUNT(*)::numeric / NULLIF(ct.total_records,0)::numeric) * 100,2) AS ratio
FROM
    return_cleaned rc
CROSS JOIN 
    customer_type ct
GROUP BY
    rc.customer_type_cleaned,
    ct.total_records
ORDER BY
    total_records__filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


customer_type_cleaned,total_records__filter,total_records,ratio
RS Swasta,27,59,45.76
Klinik,19,59,32.20
RS Pemerintah,13,59,22.03


In [51]:
%%sql

SELECT
    CASE
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
        WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
    END AS quarter,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    customer_type_cleaned ='RS Swasta'
GROUP BY
    quarter
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_records
Q4,10
Q2,8
Q3,8
Q1,1


In [59]:
%%sql

WITH 
ct_month AS (
    SELECT
        CASE
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
            WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
        END AS quarter,
        COUNT(*) AS total_records
    FROM
        return_cleaned
    GROUP BY
        quarter
),
ct_month_filter AS (
    SELECT
            CASE
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 1 AND 3 THEN 'Q1'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 4 AND 6 THEN 'Q2'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 7 AND 9 THEN 'Q3'
                WHEN EXTRACT(MONTH FROM return_date) BETWEEN 10 AND 12 THEN 'Q4'
            END AS quarter,
            COUNT(*) AS total_records_filter
    FROM
        return_cleaned
    WHERE
        customer_type_cleaned ='RS Swasta'
    GROUP BY
        quarter
)

SELECT
    ctm.quarter,
    ctmf.total_records_filter,
    ctm.total_records,
    ROUND((ctmf.total_records_filter::numeric / NULLIF(ctm.total_records,0)::numeric) * 100,2) AS ratio

FROM
    ct_month ctm
LEFT JOIN
    ct_month_filter ctmf 
        ON ctm.quarter = ctmf.quarter
ORDER BY
    ctmf.total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


quarter,total_records_filter,total_records,ratio
Q4,10,22,45.45
Q3,8,14,57.14
Q2,8,15,53.33
Q1,1,8,12.50


In [60]:
%%sql

SELECT
    kategori,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    customer_type_cleaned ='RS Swasta'
GROUP BY
    kategori
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
5 rows affected.


kategori,total_records
Bedah & Tindakan,7
Perawatan & Rawat Inap,7
Diagnostik,6
Laboratorium,4
Consumable,3


In [65]:
%%sql

SELECT
    return_reason_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    customer_type_cleaned ='RS Swasta'
GROUP BY
    return_reason_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_records
Salah Input Sistem,11
Salah Pesan Customer,10
Barang Rusak,4
Kadaluarsa,2


In [66]:
%%sql

WITH ct_kategori_filter AS
(
    SELECT
        return_reason_cleaned,
        COUNT(*) AS total_records
    FROM
        return_cleaned
    WHERE
        customer_type_cleaned ='RS Swasta'
    GROUP BY
        return_reason_cleaned
)

SELECT
    rc.return_reason_cleaned,
    ctkf.total_records AS total_records_filter,
    COUNT(*) AS total_records,
    ROUND((ctkf.total_records::numeric / NULLIF(COUNT(*),0)::numeric) * 100,2) AS ratio
FROM
    return_cleaned rc
LEFT JOIN
    ct_kategori_filter ctkf
        ON rc.return_reason_cleaned = ctkf.return_reason_cleaned
GROUP BY
    rc.return_reason_cleaned,
    total_records_filter
ORDER BY
    total_records_filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
4 rows affected.


return_reason_cleaned,total_records_filter,total_records,ratio
Salah Input Sistem,11,28,39.29
Salah Pesan Customer,10,20,50.00
Barang Rusak,4,7,57.14
Kadaluarsa,2,4,50.00


In [69]:
%%sql

SELECT
    region_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
WHERE
    customer_type_cleaned ='RS Swasta'
GROUP BY
    region_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
8 rows affected.


region_cleaned,total_records
Jakarta Barat,6
Jakarta Selatan,5
Tangerang,4
Bekasi,4
Jakarta Utara,3
Depok,2
Jakarta Timur,2
Jakarta Pusat,1


In [68]:
%%sql

SELECT *
FROM
    return_cleaned
LIMIT 3

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


retur_id,invoice_id,sku_id,customer_id,return_date,sku_name,kategori,value_return,return_reason_cleaned,customer_type_cleaned,region_cleaned,upper_bound
RT00042,010029,D001,RSP053,2024-03-02,Tensimeter Digital Lengan,Diagnostik,1468000.0,Barang Rusak,RS Pemerintah,Jakarta Utara,4107875.0
RT00046,010092,B002,RSS044,2024-05-14,Gunting Bedah Bengkok 14cm,Bedah & Tindakan,417500.0,Salah Pesan Customer,RS Swasta,Jakarta Selatan,4107875.0
RT00021,009825,L002,RSS056,2024-11-22,Tabung Vacutainer Plain,Laboratorium,478000.0,Salah Input Sistem,RS Swasta,Bekasi,4107875.0


In [70]:
%%sql

SELECT
    region_cleaned,
    COUNT(*) AS total_records
FROM
    return_cleaned
GROUP BY
    region_cleaned
ORDER BY
    total_records DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_records
Jakarta Barat,13
Bekasi,10
Jakarta Selatan,10
Jakarta Timur,6
Tangerang,5
Jakarta Utara,5
Bogor,4
Jakarta Pusat,3
Depok,3


In [ ]:
%%sql

WITH region AS
(
    SELECT
        COUNT(*) AS total_records
    FROM
        return_cleaned
)

SELECT
    rc.region_cleaned,
    COUNT(*) AS total_records__filter,
    g.total_records AS total_records,
    ROUND((COUNT(*)::numeric / NULLIF(g.total_records,0)::numeric) * 100,2) AS rate
    return_cleaned rc
CROSS JOIN 
    region g
GROUP BY
    rc.region_cleaned,
    g.total_records
ORDER BY
    total_records__filter DESC

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
9 rows affected.


region_cleaned,total_records__filter,total_records,ratio
Jakarta Barat,13,59,22.03
Bekasi,10,59,16.95
Jakarta Selatan,10,59,16.95
Jakarta Timur,6,59,10.17
Jakarta Utara,5,59,8.47
Tangerang,5,59,8.47
Bogor,4,59,6.78
Depok,3,59,5.08
Jakarta Pusat,3,59,5.08


**Conclusion:** 
 

# 4. Trend Over Time

**Conclusion:** 
 